client/components/masonry-grid/index.tsx

```diff
@@ -0,0 +1,86 @@
+import clsx from 'clsx';
+import { debounce } from 'lodash';
+import { FC, PropsWithChildren, useRef, useLayoutEffect } from 'react';
+
+import './style.scss';
+
+const calculateMasonryGrid = ( wrapper: HTMLElement ) => {
+	const wrapperStyle = getComputedStyle( wrapper );
+	const columnCount = wrapperStyle.gridTemplateColumns.split( ' ' ).length;
+	const parsedRowGap = /(\d+(\.\d+)?)px/.exec( wrapperStyle.rowGap );
+	const rowGap = parseFloat( parsedRowGap?.[ 1 ] ?? '0' );
+
+	const items = Array.from( wrapper.children ) as HTMLElement[];
+
+	if ( columnCount === 1 ) {
+		items.forEach( ( item ) => {
+			item.style.removeProperty( 'margin-top' );
+		} );
+
+		return;
+	}
+
+	// Always reset all items on the first row, since the number of grid columns is variable
+	items.slice( 0, columnCount ).forEach( ( item ) => {
+		item.style.removeProperty( 'margin-top' );
+	} );
+
+	// We calculate the difference between the top coordinates of each item with the bottom
+	// coordinates of the first item in the same column. This value is then used to set a negative
+	// `margin-top`, simulating a Masonry layout
+	items.slice( columnCount ).forEach( ( item, i ) => {
+		const aboveItemData = items[ i ].getBoundingClientRect();
+		const currentItemData = item.getBoundingClientRect();
+
+		const currentMarginTop = parseFloat( item.style.marginTop ) || 0;
+
+		const marginTop = aboveItemData.bottom - currentItemData.top + rowGap + currentMarginTop;
+
+		if ( marginTop <= -1 ) {
+			item.style.marginTop = `${ marginTop }px`;
+		} else {
+			item.style.removeProperty( 'margin-top' );
+		}
+	} );
+};
+
+type MasonryGridProps = PropsWithChildren< {
+	className?: string;
+} >;
+
+export const MasonryGrid: FC< MasonryGridProps > = ( { children, className } ) => {
+	const ref = useRef< HTMLDivElement >( null );
+
+	const debouncedCalculateMasonryGrid = debounce( ( element: HTMLElement ) => {
+		calculateMasonryGrid( element );
+	}, 100 );
```

Comment:
```suggestion
	const debouncedCalculateMasonryGrid = debounce( ( element: HTMLElement ) => {
		calculateMasonryGrid( element );
	}, 100, { leading: true, trailing: false } );
```
We currently get a flash of unstyled content. I haven't diagnosed closely, but I suspect the way we debounce this is the reason why. 

To mitigate that, we can either follow my suggestion here, or adopt my approach from the Pattern Library, where we debounce just in the space of a single animation frame

https://github.com/Automattic/wp-calypso/blob/d2f640d00a8f4b4782dca1b9153454ed3fc30c43/client/my-sites/patterns/components/pattern-gallery/client.tsx#L22-L31

I actually think the Pattern Library approach is pretty good, but we could probably go either way.